concept: benchmarking: introducing results (#229)
micro-ROS/micro-ROS.github.io: _docs/concepts/benchmarking/results/index.md

---
title: Results
redirect_from: /benchmarking/results/
permalink: /docs/concepts/benchmarking/results/
---

## Table of contents

* [From tracing to benchmarking](#from-tracing-to-benchmarking)
* [Communication results](#communication-results)
* [Real-time results](#real-time-results)
* [Executions](#executions)
* [Function execution usage](#function-execution-usage)
* [Static memory usage](#static-memory-usage)
* [Dynamic memory usage](#dynamic-memory-usage)
* [Power consumption](#power-consumption)

## From tracing to benchmarking

> The low-level RTOS was instrumented in specific way that provide different categories of benchmarking results (memory, execution, etc...). The gathered data is following the [Common Trace Format](https://diamon.org/ctf/). The Trace is later exploited using the [Babeltrace API](https://babeltrace.org/)trace manipulation toolkit. The data interpretation is up to the user.

Reviewer: ```suggestion
The low-level RTOS was instrumented in specific way that provide different categories of benchmarking results (memory, execution, etc...). The gathered data is following the [Common Trace Format](https://diamon.org/ctf/). The raw trace is later processed using the [Babeltrace API](https://babeltrace.org/) trace manipulation toolkit. The data interpretation is up to the user.
```

> More information about benchmarks results and methodologies are dealt with within the following document (TODO link to the deliverable).

Reviewer: Fix this ToDo or omit the whole sentence for now.

> The results, on which the interpretation are done, are available in the [benchmarking_results](https://github.com/micro-ROS/benchmarking-results/tree/master/aug2020) repository.

Reviewer: Please link https://github.com/micro-ROS/benchmarking on the concept page similar as you link the https://github.com/micro-ROS/benchmarking-results repo here.

Reviewer: Furthermore, can you write some sentences on the applications (its characteristic) that you used here? As far as I understood, there is not predefined set of benchmark apps in https://github.com/micro-ROS/benchmarking/, or?

> 

Reviewer: Please also describe the hardware and RTOS (here NuttX) setup used in your experiments.

[01:00:21.446789142] (+0.000010095) 0 thread_resume: { thread_id = 3 }
[01:00:21.446841333] (+0.000052191) 0 thread_suspend: { thread_id = 3 }
[01:00:21.446850857] (+0.000009524) 0 thread_resume: { thread_id = 7 }
[01:00:21.447010571] (+0.000159714) 0 thread_suspend: { thread_id = 7 }
[01:00:21.447020285] (+0.000009714) 0 thread_resume: { thread_id = 3 }
```

_Observations:_

According to the results taken above, the software is running a deterministic way. Indeed, by looking closer, it is noticeable that the running sequence is the same.

Additionally, timing deltas between correlated events have a really low variation when switch (consecutive thread_suspend/thread_resume).

Moreover, the scheduler performs fast context switches, on average the are around 10 microseconds.

## Executions

Following is depicted the execution benchmarking. The amount of time spend on the CPU split into two parts (I/O operations, Subscriber operations):

![](./images/bm_execution.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

_Observations:_
    According to the data, most of the time was spent during I/O operations.

> ## Function execution usage

Reviewer: I propose to call this _function invocations_ or _function call statistics_.

Below is the representation of the static memory analysis:

![](./images/bm_max_static_memory.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

_Observations:_

The 6LoWPAN is the medium taking the most of the static memory as this protocol is running on top of IP version 6.

## Dynamic memory usage

The graphic below is showing the total number dynamic allocations. Each bin is split into group of chunk memory blocks. For instance a block with the legend colored dark green represents all block that are greater than the previous group of chunk memory, here 16 bytes. But that the are lower or equal to 32 bytes):

![](./images/bm_allocation_nbr.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

_Observations:_

Independently on the communication mediums, blocks allocated are not huge and not numerous.
The most of the allocations are happening during initialisation.

## Power consumption

The power consumption categorised by communication medium below:

![](./images/bm_power.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

_Observations:_

> The medium has a high impact on the selection of the communication medium. This high throughput of a communication medium is coming at a price:  power consumption. The Ethernet provided a hight throughput at the price of a higher power consumption. Whereas, the serial is provide a low bitrate with the advantage to consume a lot less energy.

Reviewer: ```suggestion
The medium has a high impact on the selection of the communication medium. This high throughput of a communication medium is coming at a price:  power consumption. Ethernet provides a high throughput, but at the price of a higher power consumption. Serial provides a low bitrate with the advantage to it consumes a lot less energy.
```

Using only the timestamped measurements allows to make delta calculation offline.

The software configuration is likely to change. However the software role will be kept the same:

 * Agent running on a PC
 * Subscriber running on one Olimex STM32-E407,
 * Publisher running on one Olimex STM32-E407.

At a hardware level the USB - CDC/ACM console is going to be used on the Olimex STM32-E407 boards for Ethernet and Serial benchmarkings. For 6LoWPAN the serial USART6 will be used as the console to reduce the memory footprint and execution impact. Hence the USB OTG1 of both Olimex STM32-E407 boards shall be connected to a computer. Additional hardware setup must performed but will depend on the type of topology (Ethernet / Serial / 6LoWPAN):

The results shown below are the fruit of a same application: The **publisher**.
Only the micro-ros related functions were benchmarked. Hence, all the serial/IP/radio configuration parts of the code are benchmarked.

> Below the data processing described. From the serialised binary CTF data to the output from the Babeltrace API:

Reviewer 2: ```suggestion
Below the data processing is described, from the serialized binary CTF data to the output from the Babeltrace API:
```

![](./images/bm_dataflow.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

## Communication results

Following is the communication bitrate RX/TX:

![](./images/bm_com.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

_Observations:_

  According to the data, the Ethernet performs the best, which is expected.

## Real-time results

> Following is the execution benchmarking

Reviewer 2: ```suggestion
Below we report the execution benchmarking
```

[01:00:21.446335809] (+0.000052286) 0 thread_suspend: { thread_id = 3 }
[01:00:21.446345333] (+0.000009524) 0 thread_resume: { thread_id = 7 }
[01:00:21.446505333] (+0.000160000) 0 thread_suspend: { thread_id = 7 }
[01:00:21.446514952] (+0.000009619) 0 thread_resume: { thread_id = 3 }
[01:00:21.446564190] (+0.000049238) 0 thread_suspend: { thread_id = 3 }
[01:00:21.446573714] (+0.000009524) 0 thread_resume: { thread_id = 0 }
[01:00:21.446597714] (+0.000024000) 0 thread_suspend: { thread_id = 0 }
[01:00:21.446607714] (+0.000010000) 0 thread_resume: { thread_id = 3 }
[01:00:21.446645333] (+0.000037619) 0 thread_suspend: { thread_id = 3 }
[01:00:21.446654857] (+0.000009524) 0 thread_resume: { thread_id = 0 }
[01:00:21.446779047] (+0.000124190) 0 thread_suspend: { thread_id = 0 }
[01:00:21.446789142] (+0.000010095) 0 thread_resume: { thread_id = 3 }
[01:00:21.446841333] (+0.000052191) 0 thread_suspend: { thread_id = 3 }
[01:00:21.446850857] (+0.000009524) 0 thread_resume: { thread_id = 7 }
[01:00:21.447010571] (+0.000159714) 0 thread_suspend: { thread_id = 7 }
[01:00:21.447020285] (+0.000009714) 0 thread_resume: { thread_id = 3 }
```

_Observations:_

> According to the results taken above, the software is running a deterministic way. Indeed, by looking closer, it is noticeable that the running sequence is the same.

Reviewer 2: ```suggestion
According to the results reported above, the software is running in a deterministic way. Indeed, by looking closer, it is noticeable that the running sequence is the same.
```

> Additionally, timing deltas between correlated events have a really low variation when switch (consecutive thread_suspend/thread_resume).

Reviewer 2: ```suggestion
Additionally, timing deltas between correlated events have a really low variation when switching between events (consecutive thread_suspend/thread_resume).
```

> Moreover, the scheduler performs fast context switches, on average the are around 10 microseconds.

Reviewer 2: ```suggestion
Moreover, the scheduler performs fast context switches, which on average take around 10 microseconds.
```

## Executions

> Following is depicted the execution benchmarking. The amount of time spend on the CPU split into two parts (I/O operations, Subscriber operations):

Reviewer 2: ```suggestion
Below we report the execution benchmarking. The amount of time spent on the CPU is split into two parts (I/O operations and Subscriber operations):
```

I also want to mention that the same sentence (`Following is depicted the execution benchmarking`) appears at line 67. Possibly, one of the two is misplaced.

![](./images/bm_execution.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

_Observations:_
    According to the data, most of the time was spent during I/O operations.

## Function call statistics

> Below are depicted the function calls count per each communication medium:

Reviewer 2: ```suggestion
Below we report the function calls count per each communication medium:
```

**Ethernet**
![](./images/fusage_eth.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

**Serial**
![](./images/fusage_serial.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

**6LoWPAN**
 ![](./images/fusage_6lowpan.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

## Static memory usage

> Below is the representation of the static memory analysis:

Reviewer 2: ```suggestion
Below we provide the results for the static memory analysis:
```

![](./images/bm_max_static_memory.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

_Observations:_

> The 6LoWPAN is the medium taking the most of the static memory as this protocol is running on top of IP version 6.

Reviewer 2: ```suggestion
The 6LoWPAN is the medium consuming most of the static memory as this protocol is running on top of IP version 6.
```

## Dynamic memory usage

> The graphic below is showing the total number dynamic allocations. Each bin is split into group of chunk memory blocks. For instance a block with the legend colored dark green represents all block that are greater than the previous group of chunk memory, here 16 bytes. But that the are lower or equal to 32 bytes):

Reviewer 2: I think this paragraph is a little bit confused. Also, by looking at the graph I'm not sure that the dark green block corresponds to the numbers that are given in this sentence. Could you try to rephrase it?

![](./images/bm_allocation_nbr.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

_Observations:_

> Independently on the communication mediums, blocks allocated are not huge and not numerous.

Reviewer 2: ```suggestion
Independent of the communication medium used, the blocks allocated are not too big nor too many.
```

The most of the allocations are happening during initialisation.

## Power consumption

> The power consumption categorised by communication medium below:

Reviewer 2: ```suggestion
Below, you can find the power consumption categorized by communication medium:

Author: Thank you for your time @FranFin,
Usually i try to use the passive voice when writing document and avoid using pronominals.
Also regarding the "categorized/categorised" I was told (by your company) that we are using British English in our documents, that's why usually 'Z' are 'S'

![](./images/bm_power.png){:.img-responsive and style="max-width: 100%; margin-left: auto; margin-right: auto;"}

_Observations:_

> The medium has a high impact on the selection of the communication medium. This high throughput of a communication medium is coming at a price:  power consumption. The Ethernet provides a high throughput, but at the price of a higher power consumption. Serial communciation medium provides a low bitrate with the advantage to consuming a lot less energy.

Reviewer 2: ```suggestion
The communication medium has a high impact on the throughput that can be achieved, which in turn has an impact on power consumption. The Ethernet provides a high throughput, but at the price of higher power consumption while Serial transport provides a low bitrate with the advantage to consuming a lot less energy.
```
